SignalServiceKit/src/Messages/Interactions/TSQuotedMessage.h

```diff
@@ -2,11 +2,47 @@
 //  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
 //
 
+#import <Mantle/MTLModel.h>
 #import <SignalServiceKit/TSYapDatabaseObject.h>
 
 NS_ASSUME_NONNULL_BEGIN
 
-@interface TSQuotedMessage : TSYapDatabaseObject
+@class OWSSignalServiceProtosDataMessage;
+@class OWSSignalServiceProtosEnvelope;
+@class TSAttachment;
+@class TSAttachmentStream;
+@class TSQuotedMessage;
+@class TSThread;
+@class YapDatabaseReadWriteTransaction;
+
+@interface OWSAttachmentInfo: MTLModel
+
+@property (nonatomic, readonly, nullable) NSString *contentType;
+@property (nonatomic, readonly, nullable) NSString *sourceFilename;
+
+// This is only set when sending a new attachment so we have a way
+// to reference the original attachment when generating a thumbnail.
+// We don't want to do this until the message is saved, when the user sends
+// the message so as not to end up with an orphaned file.
+@property (nonatomic, readonly, nullable) NSString *attachmentId;
+
+// References a yet-to-be downloaded thumbnail file
+@property (atomic, nullable) NSString *thumbnailAttachmentPointerId;
+
+// References an already downloaded or locally generated thumbnail file
+@property (atomic, nullable) NSString *thumbnailAttachmentStreamId;
+
+- (instancetype)init NS_UNAVAILABLE;
+
+- (instancetype)initWithAttachmentId:(nullable NSString *)attachmentId
+                         contentType:(NSString *)contentType
+                      sourceFilename:(NSString *)sourceFilename NS_DESIGNATED_INITIALIZER;
+
+- (instancetype)initWithAttachment:(TSAttachment *)attachment;
+
+@end
+
+@interface TSQuotedMessage : MTLModel
```

Comment:
Note: TSQuotedMessage is now a MTLModel, not a TSYapDatabaseObject.

Only classes that live in their own collection should inherit from TSYapDatabaseObject.

e.g. we'd never want to call: 

```
[quotedMessage setAttachmentId:foo];
[quotedMessage saveWithTransaction:transaction]
```

Instead we do something like this:

```
[message.quotedMessage setAttachmentId:foo]`
[message saveWithTransaction:transaction];
```

Reply:
+1